OneSignalSDK/onesignal/src/main/java/com/onesignal/OSOutcomeEventsController.java

```diff
@@ -0,0 +1,324 @@
+                    return; // finish method
+            }
+        }
+
+        if (directSourceBody == null && indirectSourceBody == null && !unattributed) {
+            // Disabled for all channels
+            OneSignal.Log(OneSignal.LOG_LEVEL.VERBOSE, "Outcomes disabled for all channels");
+            return;
+        }
+
+        OSOutcomeSource source = new OSOutcomeSource(directSourceBody, indirectSourceBody);
+
+        final OSOutcomeEventParams eventParams = new OSOutcomeEventParams(name, source, weight);
+
+        OneSignalApiResponseHandler responseHandler = new OneSignalApiResponseHandler() {
+            @Override
+            public void onSuccess(String response) {
+                saveUniqueOutcome(eventParams);
+
+                // The only case where an actual success has occurred and the OutcomeEvent should be sent back
+                if (callback != null)
+                    callback.onSuccess(eventParams);
+            }
+
+            @Override
+            public void onFailure(int statusCode, String response, Throwable throwable) {
+                new Thread(new Runnable() {
+                    @Override
+                    public void run() {
+                        Thread.currentThread().setPriority(Process.THREAD_PRIORITY_BACKGROUND);
+                        // Only if we need to save and retry the outcome, then we will save the timestamp for future sending
+                        eventParams.setTimestamp(timestampSeconds);
+                        outcomeEventsFactory.getRepository().saveOutcomeEvent(eventParams);
+                    }
+                }, OS_SAVE_OUTCOMES).start();
+
+                OneSignal.onesignalLog(OneSignal.LOG_LEVEL.WARN,
+                        "Sending outcome with name: " + name + " failed with status code: " + statusCode + " and response: " + response +
+                                "\nOutcome event was cached and will be reattempted on app cold start");
+
+                // Return null within the callback to determine not a failure, but not a success in terms of the request made
+                if (callback != null)
+                    callback.onSuccess(null);
+            }
+        };
+
+        outcomeEventsFactory.getRepository().requestMeasureOutcomeEvent(appId, deviceType, eventParams, responseHandler);
+    }
+
+    private OSOutcomeSourceBody setIdsOnSourceByChannel(OSInfluence influence, OSOutcomeSourceBody sourceBody) {
+        switch (influence.getInfluenceChannel()) {
+            case IAM:
+                sourceBody.setInAppMessagesIds(influence.getIds());
+                break;
+            case NOTIFICATION:
+                sourceBody.setNotificationIds(influence.getIds());
+                break;
+        }
+
+        return sourceBody;
+    }
+
+    private List<OSInfluence> removeDisableInfluences(List<OSInfluence> influences) {
+        List<OSInfluence> influencesCopy = new ArrayList<>(influences);
```

Comment:
`influencesCopy` should be `influenceCopies`

Reply:
what about availableInfluences?

Reply:
thats fine too!